src/create_data_dictionary.py

```diff
@@ -0,0 +1,92 @@
+        self.dictionary["Message Spec"] = []
+        self.dictionary["Data Element Name"] = []
+        self.dictionary["Data Element Type"] = []
+        if os.path.isdir(bag_path):
+            bags_list = self.get_bags_recursive(bag_path)
+        else:
+            bags_list = [bag_path]
+        for bag_file in tqdm(bags_list, total=len(bags_list), desc="Bags Processed"):
+            bag = rosbag.Bag(bag_file, 'r')
+            self.populate_dictionary(bag)
+        self.save_dictionary()
+
+    def populate_dictionary(self, bag):
+        # Loop through messages in bag
+        
+        total_messages = bag.get_message_count()
+        for topic, msg, _ in tqdm(bag, total=total_messages, desc="Messages Processed", leave=False):
+            if (not len(self.topics_list) or topic in self.topics_list) and topic not in self.topics_processed:
+                self.topics_processed.add(topic)
+                msg_spec = msg._type
+                fields = []
+                for field_name in msg.__slots__:
+                    # Field name and type are in the _slot_types and __slots__ attributes
+                    field_type = msg._slot_types[msg.__slots__.index(field_name)]
+                    fields.append((field_name, field_type))
+                self.dictionary["Topic Name"].append(topic)
+                self.dictionary["Message Spec"].append(msg_spec)
+                if len(fields):
+                    self.dictionary["Data Element Name"].append(fields[0][0])
+                    self.dictionary["Data Element Type"].append(fields[0][1])
+                    for i in range(1, len(fields)):
+                        self.dictionary["Topic Name"].append(topic)
+                        self.dictionary["Message Spec"].append(msg_spec)
+                        self.dictionary["Data Element Name"].append(fields[i][0])
+                        self.dictionary["Data Element Type"].append(fields[i][1])
+                else:
+                    self.dictionary["Data Element Name"].append("")
+                    self.dictionary["Data Element Type"].append("")
+
+    def save_dictionary(self):
+        pd.DataFrame(self.dictionary).to_excel("dictionary.xlsx", index=False)
+
+
+    def get_bags_recursive(self, directory):
+        files = []
+        for root, _, filenames in os.walk(directory):
+            for filename in filenames:
+                if filename.endswith(".bag"):
+                    files.append(os.path.join(root, filename))
+        return files
+
+def main(argdict):
+    
+    DataDictionaryCreator(os.path.normpath(os.path.abspath(argdict["bag_in"])), argdict["topics_list"])
+
+    sys.exit()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Extract data for model identification")
+    parser.add_argument("bag_in", type=str, help="Bag to load, or directory of bags")
+    parser.add_argument("--topics_list", type=str, default = "", help="Path to csv file containing desired topics")
```

Comment:
I'm a little unclear on why we pass a configured list of topics to read? Instead of create a dictionary for all topics in the rosbag

Reply:
Discussed offline. This is an optional argument for only reading rosbag for a specified list topics